.gitmodules

```diff
@@ -0,0 +1,4 @@
+[submodule "modules/cudaoptflow/include/opencv2/NVIDIAOpticalFlowSDK"]
+	path = modules/cudaoptflow/include/opencv2/NVIDIAOpticalFlowSDK
+	url = https://github.com/NVIDIA/NVIDIAOpticalFlowSDK.git
```

Comment:
Git submodules are not allowed.

This significantly increases contribution barrier.

Reply:
Submodule is only for pulling NVOF SDK headers. Headers are tied to specific version of SDK and hence we thought getting them via submodule was the best option.
Can you please suggest any other alternatives? Headers don't follow opencv coding convention and SDK is already available publicly at NVIDIA devzone.

Reply:
if these headers are under appropriate license (BSD-like), then you can copy them into this repository.

> Headers are tied to specific version of SDK

This may be a problem (even with submodule).
What is the process to deliver compatible headers to users?


---

One more alternative is download and unpack sources by CMake. [Example](https://github.com/opencv/opencv_contrib/blob/3.4.0/modules/dnn_modern/CMakeLists.txt#L16-L31)
Several sources can be applied depending on users configuration (CUDA SDK version).

---

Also we need CMake checks for supported CUDA versions only (and disable functionality for unsupported versions - throw exception).

Reply:
> if these headers are under appropriate license (BSD-like), then you can copy them into this repository.
> Headers are under  BSD-3 license but they don't follow opencv coding styles and we can't change that since SDK is already public
> > Headers are tied to specific version of SDK
> So right now SDK 1.0 is public and these headers are from SDK 1.0. Submodule is linked at particular branch(nvof_1_0). So in future for SDK 2.0, we will have new branch(nvof_2_0) with SDK 2.0 with new headers and we will update the submodule dependency here.
> This may be a problem (even with submodule).
> What is the process to deliver compatible headers to users?
> Headers will be as per SDK version. Now its SDK 1.0. Later with SDK 2.0 we will update new headers.
> One more alternative is download and unpack sources by CMake. [Example](https://github.com/opencv/opencv_contrib/blob/3.4.0/modules/dnn_modern/CMakeLists.txt#L16-L31)
> Several sources can be applied depending on users configuration (CUDA SDK version).
> 
> Also we need CMake checks for supported CUDA versions only (and disable functionality for unsupported versions - throw exception).
Mostly all CUDA versions are supported. Backward compatibility is maintained. Only requirement is Turing based GPU for HW acceleration and it its not present then appropriate error message is returned.

Reply:
> >  if these headers are under appropriate license (BSD-like), then you can copy them into this repository.
> Headers are under  BSD-3 license but they don't follow opencv coding styles and we can't change that since SDK is already public

Reply:
>>Headers are tied to specific version of SDK
> So right now SDK 1.0 is public and these headers are from SDK 1.0. Submodule is linked at particular branch(nvof_1_0). So in future for SDK 2.0, we will have new branch(nvof_2_0) with SDK 2.0 with new headers and we will update the submodule dependency here.
>>This may be a problem (even with submodule).
>>What is the process to deliver compatible headers to users?
> Headers will be as per SDK version. Now its SDK 1.0. Later with SDK 2.0 we will update new headers.
>>One more alternative is download and unpack sources by CMake. [Example](https://github.com/opencv/opencv_contrib/blob/3.4.0/modules/dnn_modern/CMakeLists.txt#L16-L31)
>> Several sources can be applied depending on users configuration (CUDA SDK version).
> We felt managing headers via submodule is the best way.
>>Also we need CMake checks for supported CUDA versions only (and disable functionality for unsupported versions - throw exception).
> Mostly all CUDA versions are supported. Backward compatibility is maintained. Only requirement is Turing based GPU for HW acceleration and it its not present then appropriate error message is returned.

Reply:
You can put (copy) these headers "as is" (without modifications) - like other 3rdparty components sources.

> they don't follow opencv coding styles

This is not a problem, because these headers are internal (not exposed as OpenCV public API).

Reply:
Sure. Thanks .Let me discuss this within my team. And il update it here as soon as possible.

Reply:
Hey @alalek 
I had a discussion with my team and we decided to go with downloading via cmake. Thanks for giving a sample. I updated CMakeLists.txt in same way to fetch our headers. I force pushed so as to avoid un-necessary CL's. I hope this is fine? Can we proceed with further review?